### DISCUSSION TITLE
How does this handles large number of functions?

### DISCUSSION THREAD

**Discussion by vicary** [floydspace/serverless-esbuild#164] [General]
I am new to esbuild, so this discussion maybe in a grey area between `serverless-esbuild` and `esbuild`.

I have an AppSync project in production that contains 300-ish lambdas resolvers in TypeScript, still growing.

With webpack I have to throttle the concurrency with `thread-loader`, `fork-ts-checker-webpack-plugin`, `fork-ts-checker-webpack-plugin-limiter` and my own PR https://github.com/serverless-heaven/serverless-webpack/pull/681 to modify webpack a little bit in order to work.

I believe 300 entries is not a normal use case for any kind of bundlers, but AppSync just creates such an extreme play.

Do you think serverless-esbuild is able to handle 300 entries out of the box?

If not, can it handle them with plugins and/or configurations?

**Comment by olup** [comment]
No idea, but I'd be interested in the results if you try serverless-esbuild
with your setup.

Le jeu. 15 juil. 2021 à 16:31, Vicary A. ***@***.***> a
écrit :

**Reply by vicary** [reply]
You may see my result in the comment below https://github.com/floydspace/serverless-esbuild/discussions/164#discussioncomment-1012021

**Comment by floydspace** [comment]
I think `esbuild` itself is capable to handle bundling 300+ entries at the same time, because fast bundling is in the blood of esbuld. But I'm not sure if our external modules packaging logic would handle it. I would definitely suggest trying the`serverless-esbuild` with or without packaging externals. Really curious about the results.

**Reply by vicary** [reply]
The biggest problem is if it even builds. My experience in webpack tells me that it may break at multiple points, for example it very quickly exceeded JS heap size and the limit of open file descriptors. I'll definitely test it out though.

Do you by any chance have any esbuild plugins in mind if I ever need one?

**Reply by floydspace** [reply]
I don't think you need any other plugins, what I would suggest though is to put your `node_modules` in shared lambda layer and disable packing external modules, in this way you at least eliminate overflow with file system interaction

**Reply by vicary** [reply]
Sorry but I'm not really sure how to force an exclusion on all `node_modules`, shall I make use of the `packagePath` option to a fake `package.json` without any `dependencies`?

**Reply by floydspace** [reply]
so default `serverless-esbuild` configuration is to bundle everything using `esbuild`, but if you do not want to bundle dependencies from `node_modules` and you do not want them packed as externals, you need to list them in the `exclude` options, like this:
```
custom:
  esbuild:
    exclude:
      - lodash
```
in your particular case you have to exclude everything. 
But we have an open [PR](https://github.com/floydspace/serverless-esbuild/pull/119) on review which would cover your case imo. I wrote some comments there to improve it, but it was not finished.

**Reply by vicary** [reply]
jeesh motherly love of god

In my webpack setup it was 50-80 seconds per entry (with type checking), with esbuild it is 2 secs per entry. Looks like it's made with concurrencies in mind, the resulting time of the whole build process reduced from 30 mins to 3.5 mins!

If anything else it's the type checking I have to take care of, probably hook into my test pipelines and call it a day.

I have to admit this is a very promising result!

**Reply by floydspace** [reply]
Happy to hear it works, do not hesitate to make a PR to `serverless-esbuild`, any contribution is appreciated.

**Reply by vamche** [reply]
> Sorry but I'm not really sure how to force an exclusion on all `node_modules`, shall I make use of the `packagePath` option to a fake `package.json` without any `dependencies`?

FYI, you can use `esbuild.packagerOptions.scripts` to exclude `node_modules` from the final artefact. 

``` yaml
  packagerOptions:
      scripts:
        - rm -rf node_modules
```

**Reply by floydspace** [reply]
yes, but it's better just not create the `node_modules` at all, otherwise, it's just wasting time

**Reply by vicary** [reply]
I was using `@esbuild-plugins/node-resolve` to exclude it.

Tried not using that to see final bundle size, esbuild simply refuse to run. It hangs for 2 minutes then crash with the message "Service is stopped."

I guess it is about exceeding heap size limit but there is no internal stack trace for me.

**Comment by floydspace** [comment]
@vicary Also, as a side suggestion, you could try using [esbuild-loader](https://github.com/privatenumber/esbuild-loader) with your `webpack` setup and see how it handles your project. it must increase your lambdas bundling speed

**Reply by vicary** [reply]
I actually tried this one. It halves the build time for each entry for my other smaller projects, with around 50 lambdas on average. But, sadly, for the big project with 300 entries it makes no difference.
